### ISSUE TITLE
关于search任务 αKL设置和复现实验结果相关

### ISSUE THREAD

**Issue by mingyueqingfeng2023-collab** [ZJU-REAL/SDAR#14]
你好，感谢开源工作！
现有2个问题想要请教：
1.原论文在search任务中 提到αKL设置为0.01（23页 C Hyperparameters），而在SDAR\examples\sdar_trainer\run_search_3b.sh 中actor_rollout_ref.actor.kl_loss_coef=0.001，请问训练的αKL值应是？
2.在search任务上，复现Qwen2.5-3B-Instruct+SDAR的结果与原论文有差异（actor_rollout_ref.actor.kl_loss_coef=0.001，step=150，都依据原论文配置）：
Method | NQ | Triv | Pop | Hotp | 2Wk | MuS | Bam | Avg
SDAR    | 40.5 | 56.1 | 37.3 | 38.7 | 35.1 | 16.2 | 33.6 | 36.8
主要体现在Pop(37.3 vs 44.3)和 Bam(33.6 vs 66.1)，请问可能的原因是αKL的选取，还是训练不稳定需要重训几次，还是延长训练step，还是其他？

**Comment by lll6gg** [comment]
您好
1. αKL是脚本中的sdar_coef
2. 可以提供训练参数，以及训练曲线吗

**Comment by mingyueqingfeng2023-collab** [comment]
感谢回答！
1.原论文23页中：
Method | η | G    | ε |    λ |    β |  αKL |
GRPO | 1e-6 | 8 | 0.2 | — | —  | 0.01 | 
RLSD | 1e-6 | 8 | 0.2 | 0.5 | —  | 0.01 | 
SDAR | 1e-6 | 8 | 0.2 | 0.01 | 5.0 | 0.01 | 
λ=蒸馏损失系数；β=sigmoid gate sharpness；αKL=朝参考策略的KL 惩罚系数；这三者应该分别对应脚本中的sdar_coef=0.01；gate_beta=5.0；actor_rollout_ref.actor.kl_loss_coef=0.001 ；所以αKL 应该是actor_rollout_ref.actor.kl_loss_coef？（在\run_alfworld_3b.sh和run_webshop_3b.sh中actor_rollout_ref.actor.kl_loss_coef都为0.01；且在我复现原始search-r1时（官方代码仓），其原始配置中 αKL=0.001，很容易在150step左右出现actor/ppo_kl 激增，从而导致模型输出乱码，训练崩溃，通过将αKL设置为0.01，限制训练的模型不能偏移原始参考模型过远便能解决这一问题 ）

2.以下是我复现过程中的训练曲线：

<img width="3378" height="8596" alt="Image" src="https://github.com/user-attachments/assets/c57d7a49-d623-45ef-93d4-89fc146b10e2" />

<img width="3344" height="2821" alt="Image" src="https://github.com/user-attachments/assets/b91a23d7-4582-4555-98dc-1b3795b23b2a" />

注：在50step我停下来添加了抽样代码，以便能在log日志看到rollout过程中的部分轨迹，以分析模型学习的过程；而后我在50step断点处继续训练


训练参数如下（训练GPU配置为4*NVIDIA A100-SXM4-80GB）：
set -x
ENGINE=${1:-vllm}

# export WANDB_API_KEY=your_key_here
export HIGHLIGHT_CONFIGS='<search>:0,0,255;</search>:0,0,255;<information>:255,0,0;</information>:255,0,0'
export PRINT_ROLLOUT_NUM=4
export PRINT_ROLLOUT_PROB=1.0
# SDAR hyperparameters
sdar_coef=0.01
gate_beta=5.0
skill_all=false

experiment_name="20260529_1453_sdar_qwen2.5_3b_coef${sdar_coef}_beta${gate_beta}_skillall${skill_all}"

train_data_size=128
# train_data_size=64
val_data_size=512
group_size=8

TRAIN_DATA="/Search-r1/Search-R1/data/searchR1_processed_direct/train.parquet"
VAL_DATA="/Search-r1/Search-R1/data/searchR1_processed_direct/test.parquet"


python3 -m verl.trainer.main_sdar \
    algorithm.adv_estimator=grpo \
    data.train_files=$TRAIN_DATA \
    data.val_files=$VAL_DATA \
    data.train_batch_size=$train_data_size \
    data.val_batch_size=$val_data_size \
    data.max_prompt_length=4096 \
    data.max_response_length=512 \
    data.filter_overlong_prompts=True \
    data.truncation='left' \
    data.return_raw_chat=True \
    actor_rollout_ref.model.path=/Search-r1/Search-R1/models/Qwen2.5-3B-Instruct \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.optim.lr_warmup_steps_ratio=0.1 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=256 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=32 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=$ENGINE \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.5 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=32 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.use_invalid_action_penalty=True \
    actor_rollout_ref.actor.invalid_action_penalty_coef=0.01 \
    algorithm.use_kl_in_reward=False \
    +algorithm.sdar.sdar_coef=$sdar_coef \
    +algorithm.sdar.gate_beta=$gate_beta \
    +algorithm.sdar.skills_dir=skills/search \
    +algorithm.sdar.skill_all=$skill_all \
    env.env_name=search \
    env.seed=0 \
    env.max_steps=4 \
    env.rollout.n=$group_size \
    env.history_length=4 \
    env.search.search_url='http://0.0.0.0:8000/retrieve' \
    trainer.critic_warmup=0 \
    trainer.logger=['console','wandb'] \
    trainer.project_name='verl_agent_search' \
    trainer.experiment_name=$experiment_name \
    trainer.n_gpus_per_node=4 \
    trainer.ray_wait_register_center_timeout=600 \
    trainer.nnodes=1 \
    trainer.save_freq=50 \
    trainer.test_freq=150 \
    trainer.total_training_steps=150 \
    trainer.val_before_train=False $@

**Comment by gitrxh** [comment]
@mingyueqingfeng2023-collab 您好，请教一下，请问您换了参数复现出这个search上的效果了吗？
